Public/Src/Cache/ContentStore/Library/Service/Grpc/GrpcCopyClient.cs

```diff
@@ -83,74 +87,166 @@ public async Task<BoolResult> ShutdownAsync(Context context)
+        /// </summary>
+        public async Task<CopyFileResult> CopyToAsync(Context context, ContentHash hash, Func<Stream> streamFactory, CancellationToken ct = default(CancellationToken))
+        {
             try
             {
-                AsyncServerStreamingCall<CopyFileResponse> response = _client.CopyFile(new CopyFileRequest
+                CopyFileRequest request = new CopyFileRequest()
                 {
                     TraceId = context.Id.ToString(),
-                    HashType = (int)contentHash.HashType,
-                    ContentHash = contentHash.ToByteString(),
-                    // TODO: If `Drive` is expected to be the drive of the file on the source machine, then this should have nothing to do with the destination's drive
-                    Drive = "B",
+                    HashType = (int) hash.HashType,
+                    ContentHash = hash.ToByteString(),
+                    Drive = "X",
                     Offset = 0,
-                    Compression = compression
-                });
+                    Compression = SupportsCompression ? CopyCompression.Gzip : CopyCompression.None
+                };
+
+                AsyncServerStreamingCall<CopyFileResponse> response = _client.CopyFile(request);
+
+                Metadata headers = await response.ResponseHeadersAsync.ConfigureAwait(false);
+
+                // If the remote machine couldn't be contacted, GRPC returns an empty
+                // header collection. GRPC would throw an RpcException when we tried
+                // to stream response, but by that time we would have created target
+                // stream. To avoid that, exit early instead.
+                if (headers.Count == 0)
+                {
+                    return new CopyFileResult(CopyFileResult.ResultCode.SourcePathError, $"Failed to connect to copy server {_host} at port {_grpcPort}.");
+                }
 
-                bytesReceived = await StreamContentAsync(stream, response.ResponseStream);
+                // Check header collection for server-side errors.
+                string exception = null;
+                string message = null;
+                CopyCompression compression = CopyCompression.None;
+                foreach(Metadata.Entry header in headers)
+                {
+                    if (header.Key == "exception")
+                    {
+                        exception = header.Value;
+                    }
+                    if (header.Key == "message")
+                    {
+                        message = header.Value;
+                    }
+                    if (header.Key == "compression")
+                    {
+                        Enum.TryParse<CopyCompression>(header.Value, out compression);
+                    }
+                }
+                if (exception != null)
+                {
+                    Debug.Assert(message != null);

```

Comment:
Switch to Contract?

Reply:
Boo, hiss. I want a system that allows me to do asserts in a debug build that are not checked in retail.